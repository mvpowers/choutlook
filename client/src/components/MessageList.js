import React, { Component } from 'react';
import styled from 'styled-components';
import PropTypes from 'prop-types';
import { FaChevronDown } from 'react-icons/lib/fa';
import { Message } from './';

const uuidv4 = require('uuid/v4');

const Container = styled.div`
  background-color: #fff;
  border: 1px solid #eaeaea;
  width: 350px;
  overflow-y: scroll;
`;

const Header = styled.div`
  display: flex;
  flex-direction: row;
  justify-content: space-between;
  padding: 10px 15px;
  align-items: center;
  border-bottom: 1px solid #eaeaea;
`;

const Title = styled.div`
  font-size: 17px;
  margin-left: 5px;
`;

const Filter = styled.div`
  display: flex;
  flex-direction: row;
  font-size: 12px;
  color: #077bd8;
  align-items: center;
`;

const Text = styled.div`
  padding: 15px;
`;

const Period = styled.div`
  background-color: #f8f8f8;
  border-bottom: 1px solid #eaeaea;
  padding: 10px;
  font-size: 14px;
`;

export default class MessageList extends Component {
  componentDidUpdate() {
    if (this.props.replyFocus) {
      this.scrollToBottom();
    }
  }

  scrollToBottom() {
    this.el.scrollIntoView({ behavior: 'smooth' });
  }

  render() {
    const { displayMsg } = this.props;
    return (
      <Container>
        <Header>
          <Title>Inbox</Title>
          <Filter>
            <Text>Filter</Text>
            <FaChevronDown size={12} />
          </Filter>
        </Header>
        <Period>Today</Period>
        {displayMsg.map(msg => (
          <Message key={uuidv4()} user={msg.user} msg={msg.message} />
        ))}
        <div
          ref={el => {
            this.el = el;
          }}
        />
      </Container>
    );
  }
}

MessageList.propTypes = {
  displayMsg: PropTypes.arrayOf(
    PropTypes.shape({
      time: PropTypes.string.isRequired,
      user: PropTypes.string.isRequired,
      message: PropTypes.string.isRequired,
    })
  ).isRequired,
  replyFocus: PropTypes.bool.isRequired,
};
