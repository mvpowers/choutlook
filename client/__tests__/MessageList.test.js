import React from 'react';
import Enzyme, { shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import MessageList from '../src/components/MessageList';

Enzyme.configure({ adapter: new Adapter() });

describe('MessageList', () => {
  it('should be defined', () => {
    expect(MessageList).toBeDefined();
  });
  it('should render correctly', () => {
    const tree = shallow(
      <MessageList
        displayMsg={[{ time: 'time', user: 'user', message: 'message' }]}
        replyFocus={false}
      />
    );
    expect(tree).toMatchSnapshot();
  });
});
