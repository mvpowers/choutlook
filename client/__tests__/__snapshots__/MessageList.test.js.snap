// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MessageList should render correctly 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <MessageList
    displayMsg={
      Array [
        Object {
          "message": "message",
          "time": "time",
          "user": "user",
        },
      ]
    }
    replyFocus={false}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "children": Array [
        <MessageList__Header>
          <MessageList__Title>
            Inbox
          </MessageList__Title>
          <MessageList__Filter>
            <MessageList__Text>
              Filter
            </MessageList__Text>
            <FaChevronDown
              size={12}
            />
          </MessageList__Filter>
        </MessageList__Header>,
        <MessageList__Period>
          Today
        </MessageList__Period>,
        Array [
          <Message
            msg="message"
            user="user"
          />,
        ],
        <div />,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "children": Array [
            <MessageList__Title>
              Inbox
            </MessageList__Title>,
            <MessageList__Filter>
              <MessageList__Text>
                Filter
              </MessageList__Text>
              <FaChevronDown
                size={12}
              />
            </MessageList__Filter>,
          ],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "children": "Inbox",
            },
            "ref": null,
            "rendered": "Inbox",
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "children": Array [
                <MessageList__Text>
                  Filter
                </MessageList__Text>,
                <FaChevronDown
                  size={12}
                />,
              ],
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "class",
                "props": Object {
                  "children": "Filter",
                },
                "ref": null,
                "rendered": "Filter",
                "type": [Function],
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "function",
                "props": Object {
                  "size": 12,
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
            ],
            "type": [Function],
          },
        ],
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "children": "Today",
        },
        "ref": null,
        "rendered": "Today",
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": "aabd44de-28b1-41cb-9806-da0ee055958d",
        "nodeType": "function",
        "props": Object {
          "msg": "message",
          "user": "user",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {},
        "ref": [Function],
        "rendered": null,
        "type": "div",
      },
    ],
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": Array [
          <MessageList__Header>
            <MessageList__Title>
              Inbox
            </MessageList__Title>
            <MessageList__Filter>
              <MessageList__Text>
                Filter
              </MessageList__Text>
              <FaChevronDown
                size={12}
              />
            </MessageList__Filter>
          </MessageList__Header>,
          <MessageList__Period>
            Today
          </MessageList__Period>,
          Array [
            <Message
              msg="message"
              user="user"
            />,
          ],
          <div />,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": Array [
              <MessageList__Title>
                Inbox
              </MessageList__Title>,
              <MessageList__Filter>
                <MessageList__Text>
                  Filter
                </MessageList__Text>
                <FaChevronDown
                  size={12}
                />
              </MessageList__Filter>,
            ],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "children": "Inbox",
              },
              "ref": null,
              "rendered": "Inbox",
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "children": Array [
                  <MessageList__Text>
                    Filter
                  </MessageList__Text>,
                  <FaChevronDown
                    size={12}
                  />,
                ],
              },
              "ref": null,
              "rendered": Array [
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "class",
                  "props": Object {
                    "children": "Filter",
                  },
                  "ref": null,
                  "rendered": "Filter",
                  "type": [Function],
                },
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "function",
                  "props": Object {
                    "size": 12,
                  },
                  "ref": null,
                  "rendered": null,
                  "type": [Function],
                },
              ],
              "type": [Function],
            },
          ],
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": "Today",
          },
          "ref": null,
          "rendered": "Today",
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": "aabd44de-28b1-41cb-9806-da0ee055958d",
          "nodeType": "function",
          "props": Object {
            "msg": "message",
            "user": "user",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {},
          "ref": [Function],
          "rendered": null,
          "type": "div",
        },
      ],
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
